compparallel.go

```diff
@@ -326,115 +325,131 @@ func getChannelOrErrorParallel[T any](
 	f func(context.Context, routing.Routing) (<-chan T, error),
 	shouldStop func() bool,
 ) (chan T, error) {
-	outCh := make(chan T)
-	errCh := make(chan error)
-	var wg sync.WaitGroup
+	var ready sync.Mutex
+	ready.Lock()
+	var outCh chan T
+	var errorsLk sync.Mutex
+	// nil errors indicate sucess
+	errors := []error{}
+
 	ctx, cancelAll := context.WithCancel(ctx)
+	fwg := jsync.NewFWaitGroup(func() {
+		if outCh != nil {
+			close(outCh)
+		} else {
+			ready.Unlock()
+		}
+
+		cancelAll()
+	}, uint64(len(routers)))
+
+	var blocking atomic.Uint64
+	blocking.Add(1) // start at one so we don't cancel while dispatching
+	var sent atomic.Bool
+
 	for _, r := range routers {
-		wg.Add(1)
+		if !r.DoNotWaitForStreamingResponses {
+			blocking.Add(1)
+		}
+
 		go func(r *ParallelRouter) {
-			defer wg.Done()
-			log.Debug("getChannelOrErrorParallel: starting execution for router ", r.Router,
-				" with timeout ", r.Timeout,
-				" and ignore errors ", r.IgnoreError,
-			)
-			tim := time.NewTimer(r.ExecuteAfter)
-			defer tim.Stop()
-			select {
-			case <-ctx.Done():
-				log.Debug("getChannelOrErrorParallel: stopping execution on router on context done for router ", r.Router,
-					" with timeout ", r.Timeout,
-					" and ignore errors ", r.IgnoreError,
-				)
-				return
-			case <-tim.C:
-				ctx, cancel := withCancelAndOptionalTimeout(ctx, r.Timeout)
-				defer cancel()
+			defer fwg.Done()
+			defer func() {
+				var remainingBlockers uint64
+				if r.DoNotWaitForStreamingResponses {
+					remainingBlockers = blocking.Load()
+				} else {
+					var minusOne uint64
+					minusOne--
```

Comment:
Why are we doing this with uints instead of ints?

Reply:
Because it's not negative.